lasagne/objectives.py

```diff
@@ -1,35 +1,50 @@
+import theano
 import theano.tensor as T
 
 
-def mse(x, t):
+def mse(x, t, m):
     """Calculates the MSE mean across all dimensions, i.e. feature
      dimension AND minibatch dimension.
 
     :parameters:
         - x : predicted values
         - t : target values
+        - m : mask; None for no mask, or a 1-dimensional array, that
+            selects which rows in `x` and `t` should count towards the
+            result
 
     :returns:
         - output : the mean square error across all dimensions
     """
-    return T.mean((x - t) ** 2)
+    if m is None:
+        return T.mean((x - t) ** 2)
+    else:
+        recip_mask_sum = T.cast(1.0 / T.sum(m), theano.config.floatX)
+        return T.sum(m * (x - t) ** 2) * recip_mask_sum
```

Comment:
I'm not sure about the renormalization. When you have some unmasked minibatches and then some half-masked minibatches, each datapoint in a half-masked minibatch will contribute twice as strongly to the weight updates as a datapoint in an unmasked minibatch. If at all, the renormalization should be optional.